From 181bb5966a93aa9a61fd76dc7373519a7c4ff4cd Mon Sep 17 00:00:00 2001
From: Eric Lundby <Eric.Lundby@gmail.com>
Date: Tue, 28 Oct 2025 09:48:22 -0600
Subject: [PATCH 2/2] 23

ERROR: Flag 'leave_barriers_on_recoverable_agent_restart' was defined more than once but with differing types. Defined in files 'external/local_xla/xla/tsl/distributed_runtime/coordination/coordination_service.cc' and 'external/local_xla/xla/tsl/distributed_runtime/coordination/coordination_service.cc'.
---
 tensorflow/core/distributed_runtime/rpc/BUILD |  2 +-
 third_party/xla/xla/service/BUILD             |  2 +-
 .../distributed_runtime/coordination/BUILD    | 42 ++++++++++++++++++-
 3 files changed, 43 insertions(+), 3 deletions(-)

diff --git a/tensorflow/core/distributed_runtime/rpc/BUILD b/tensorflow/core/distributed_runtime/rpc/BUILD
index c0351bbe448..cb454aa38b5 100644
--- a/tensorflow/core/distributed_runtime/rpc/BUILD
+++ b/tensorflow/core/distributed_runtime/rpc/BUILD
@@ -335,7 +335,7 @@ cc_library(
         "//tensorflow/core/nccl:collective_communicator",
         "//tensorflow/core/profiler/rpc:profiler_service_impl",
         "@com_google_absl//absl/strings",
-        "@local_xla//xla/tsl/distributed_runtime/coordination:coordination_service",
+        "@local_xla//xla/tsl/distributed_runtime/coordination:coordination_service_hdrs_only",
         "@local_xla//xla/tsl/distributed_runtime/rpc:async_service_interface",
     ] + tf_protos_profiler_service() + tf_grpc_dependencies() + tf_grpc_cc_dependencies(),
     alwayslink = 1,
diff --git a/third_party/xla/xla/service/BUILD b/third_party/xla/xla/service/BUILD
index ce859c54227..d735e675cad 100644
--- a/third_party/xla/xla/service/BUILD
+++ b/third_party/xla/xla/service/BUILD
@@ -5279,7 +5279,7 @@ cc_library(
     hdrs = ["global_device_id.h"],
     deps = [
         "//xla/runtime:device_id",
-        "//xla/tsl/distributed_runtime/coordination:coordination_service",
+        "//xla/tsl/distributed_runtime/coordination:coordination_service_hdrs_only",
         "@com_google_absl//absl/strings",
         "@com_google_absl//absl/types:span",
     ],
diff --git a/third_party/xla/xla/tsl/distributed_runtime/coordination/BUILD b/third_party/xla/xla/tsl/distributed_runtime/coordination/BUILD
index 962640f01dd..73df692198f 100644
--- a/third_party/xla/xla/tsl/distributed_runtime/coordination/BUILD
+++ b/third_party/xla/xla/tsl/distributed_runtime/coordination/BUILD
@@ -50,7 +50,7 @@ cc_library(
 )
 
 cc_library(
-    name = "coordination_service",
+    name = "coordination_service_impl",
     srcs = ["coordination_service.cc"],
     hdrs = ["coordination_service.h"],
     deps = [
@@ -84,6 +84,46 @@ cc_library(
     ],
 )
 
+cc_library(
+    name = "coordination_service",
+    deps = [":coordination_service_impl"],
+    visibility = ["//visibility:public"],
+)
+
+cc_library(
+    name = "coordination_service_hdrs_only",
+    hdrs = ["coordination_service.h"],
+    deps = [
+        ":coordination_client",
+        ":coordination_service_error_util",
+        ":key_value_store",
+        "//xla/tsl/distributed_runtime:call_options",
+        "//xla/tsl/lib/gtl:int_type",
+        "//xla/tsl/platform:env",
+        "//xla/tsl/platform:status",
+        "//xla/tsl/protobuf:coordination_config_proto_cc",
+        "//xla/tsl/protobuf:coordination_service_proto_cc",
+        "//xla/tsl/util:device_name_utils",
+        "@com_google_absl//absl/algorithm:container",
+        "@com_google_absl//absl/base:core_headers",
+        "@com_google_absl//absl/container:flat_hash_map",
+        "@com_google_absl//absl/container:flat_hash_set",
+        "@com_google_absl//absl/hash",
+        "@com_google_absl//absl/log",
+        "@com_google_absl//absl/log:check",
+        "@com_google_absl//absl/status",
+        "@com_google_absl//absl/status:statusor",
+        "@com_google_absl//absl/strings",
+        "@com_google_absl//absl/strings:str_format",
+        "@com_google_absl//absl/strings:string_view",
+        "@com_google_absl//absl/synchronization",
+        "@com_google_absl//absl/time",
+        "@com_google_absl//absl/types:span",
+        "@local_tsl//tsl/platform:random",
+    ],
+    visibility = ["//visibility:public"],
+)
+
 tf_proto_library(
     name = "test_device_proto",
     testonly = 1,
-- 
2.50.1 (Apple Git-155)

